libbeat/outputs/shipper/shipper_test.go

```diff
@@ -93,6 +102,236 @@ func TestConvertMapStr(t *testing.T) {
+		},
+	}
+
+	cases := []struct {
+		name       string
+		events     []beat.Event
+		expSignals []outest.BatchSignal
+		qSize      int
+	}{
+		{
+			name:   "sends a batch",
+			events: events[:1],
+			expSignals: []outest.BatchSignal{
+				{
+					Tag: outest.BatchACK,
+				},
+			},
+			qSize: 5,
+		},
+		{
+			name:   "retries not accepted events",
+			events: events,
+			expSignals: []outest.BatchSignal{
+				{
+					Tag:    outest.BatchRetryEvents,
+					Events: toPublisherEvents(events[1:]),
+				},
+			},
+			qSize: 1,
+		},
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+			defer cancel()
+			producer := sc.NewProducerMock(tc.qSize)
+			grpcServer := grpc.NewServer()
+			sc.RegisterProducerServer(grpcServer, producer)
+
+			listener, err := net.Listen("tcp", "localhost:0") // random available port
+			require.NoError(t, err)
+
+			defer grpcServer.Stop()
+			go func() {
+				_ = grpcServer.Serve(listener)
+			}()
+
+			cfg, err := config.NewConfigFrom(map[string]interface{}{
+				"server": listener.Addr().String(),
+			})
+			require.NoError(t, err)
+
+			group, err := makeShipper(
+				nil,
+				beat.Info{Beat: "libbeat", IndexPrefix: "testbeat"},
+				outputs.NewNilObserver(),
+				cfg,
+			)
+			require.NoError(t, err)
+			require.Len(t, group.Clients, 1)
+
+			batch := outest.NewBatch(tc.events...)
+			err = group.Clients[0].Publish(ctx, batch)
+			require.NoError(t, err)
+
+			require.Equal(t, tc.expSignals, batch.Signals)
+		})
+	}
+
+	t.Run("cancel the batch when the server is not available", func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+		defer cancel()
+
+		producer := sc.NewProducerMock(5)
+		grpcServer := grpc.NewServer()
+		sc.RegisterProducerServer(grpcServer, producer)
+
+		listener, err := net.Listen("tcp", "localhost:0") // random available port
+		require.NoError(t, err)
+
+		defer grpcServer.Stop()
+		// run with a 2 second delay
+		go func() {
+			<-time.After(2 * time.Second)
+			_ = grpcServer.Serve(listener)
+		}()
+
+		cfg, err := config.NewConfigFrom(map[string]interface{}{
+			"server":  listener.Addr().String(),
+			"timeout": 1, // 1 sec
+		})
+		require.NoError(t, err)
+
+		group, err := makeShipper(
+			nil,
+			beat.Info{Beat: "libbeat", IndexPrefix: "testbeat"},
+			outputs.NewNilObserver(),
+			cfg,
+		)
+		require.NoError(t, err)
+		require.Len(t, group.Clients, 1)
+
+		batch := outest.NewBatch(events...)
+
+		// try to publish without the server running
+		err = group.Clients[0].Publish(ctx, batch)
+		require.NoError(t, err)
+
+		expSignals := []outest.BatchSignal{
+			{
+				Tag: outest.BatchCancelled, // "cancelled" means there will be a retry without decreasing the TTL
+			},
+		}
+		require.Equal(t, expSignals, batch.Signals)
+
+		<-time.After(2 * time.Second)
```

Comment:
What is this delay for, can we get rid of it?

Reply:
Not really.

The algorithm of the test is:

1. Try to connect/publish to the server while it's still not available (hence the 2 sec delay before the server starts on line 286 and `cfg.timeout = 1 sec` for the publisher).
2. Then wait 2 seconds (to make sure the delayed server had enough time to eventually start)
3. Try connect/publish again and succeed this time in the bounds for the 1 sec timeout (that we set in the config).

Since there is no way for the server to notify when it starts, we have to use the delays but I can't see how they could be flaky, since these delays and timeouts are quite stable.